pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+)
+
+const (
+	controllerName = "extension.extensions.crossplane.io"
+
+	reconcileTimeout      = 1 * time.Minute
+	requeueAfterOnSuccess = 10 * time.Second
+
+	reasonCreatingRBAC       = "failed to create RBAC"
+	reasonCreatingDeployment = "failed to create deployment"
+)
+
+var (
+	log              = logging.Logger.WithName(controllerName)
+	resultRequeue    = reconcile.Result{Requeue: true}
+	requeueOnSuccess = reconcile.Result{RequeueAfter: requeueAfterOnSuccess}
+)
+
+// Reconciler reconciles a Instance object
+type Reconciler struct {
+	kube client.Client
+	factory
+}
+
+// Add creates a new Extension Controller and adds it to the Manager with default RBAC.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	r := &Reconciler{
+		kube:    mgr.GetClient(),
+		factory: &extensionHandlerFactory{},
+	}
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Extension
+	err = c.Watch(&source.Kind{Type: &v1alpha1.Extension{}}, &controllerHandler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile reads that state of the Extension for a Instance object and makes changes based on the state read
+// and what is in the Instance.Spec
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("reconciling", "kind", v1alpha1.ExtensionKindAPIVersion, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	// fetch the CRD instance
+	i := &v1alpha1.Extension{}
+	if err := r.kube.Get(ctx, req.NamespacedName, i); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	handler := r.factory.newHandler(ctx, i, r.kube)
+
+	return handler.sync(ctx)
+}
+
+type handler interface {
+	sync(context.Context) (reconcile.Result, error)
+	create(context.Context) (reconcile.Result, error)
+	update(context.Context) (reconcile.Result, error)
+}
+
+type extensionHandler struct {
+	kube client.Client
+	i    *v1alpha1.Extension
+}
+
+type factory interface {
+	newHandler(context.Context, *v1alpha1.Extension, client.Client) handler
+}
+
+type extensionHandlerFactory struct{}
+
+func (f *extensionHandlerFactory) newHandler(ctx context.Context, i *v1alpha1.Extension, kube client.Client) handler {
+	return &extensionHandler{
+		kube: kube,
+		i:    i,
+	}
+}
+
+// ************************************************************************************************
+// Syncing/Creating functions
+// ************************************************************************************************
+func (h *extensionHandler) sync(ctx context.Context) (reconcile.Result, error) {
+	if h.i.Status.ControllerRef == nil {
+		return h.create(ctx)
+	}
+
+	return h.update(ctx)
```

Comment:
Since update is "noop" - then there is no way to perform any object modification if/when `Status.ControllerRef` is set. I.e. the only way to proceed is either:
a) delete and recreate the app
b) manually patch app removing `Status.ControllerRef` to force `create` execution flow.
Perhaps this is another good example of using "observability" vs. "status field"

Reply:
This will be captured along with responding to changes to child artifacts